### ISSUE TITLE
Timestamp behavior, not working when editing

### ISSUE THREAD

**Issue by unclezoot** [cakephp/cakephp#9699]
This is a (multiple allowed):
* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.3.7
* Platform and Target:  MAMP PRO 4.0.5, PHP 5.6.25, Postgres 8

### What you did
Baked a plugin model then added Timestamp behavior as follows:
~~~~
    public function initialize(array $options)
    {
        parent::initialize($options);

        $this->addBehavior('Timestamp',[
            'events' => [
                'Model.beforeSave' => [
                    'created' => 'new',
                    'updated' => 'always'
                ]
            ]
        ]);
~~~~

Then used the baked controller to test add and edit.

### What happened
Add 'created' and 'updated' fields in the DB as expected.
Edit, however, did not alter the 'updated' field at all. Neither did it get filled after manually setting the field to NULL in the DB and repeating the edit.

### Expected Behavior
Expected the updated field to reflect the current time and at least get populated.
From looking at \ORM\Table::save, it appears to be something to do with the dirty() function not getting flagged when editing when it should be.

e.g. commenting out these lines (1485-1487):
~~~~
        if ($entity->isNew() === false && !$entity->dirty()) {
            return $entity;
        }
~~~~
appears to resolve the issue?

**Comment by markstory** [comment]
How are you updating your entities? What makes you think that `dirty()` is the problem?

**Comment by unclezoot** [comment]
Updating via the following baked controller code:

    /**
     * Edit method
     *
     * @param string|null $id Target id.
     * @return \Cake\Network\Response|void Redirects on successful edit, renders view otherwise.
     * @throws \Cake\Network\Exception\NotFoundException When record not found.
     */
    public function edit($id = null)
    {
        $target = $this->Targets->get($id, [
            'contain' => []
        ]);
        if ($this->request->is(['patch', 'post', 'put'])) {
            $target = $this->Targets->patchEntity($target, $this->request->data);
            if ($this->Targets->save($target)) {
                $this->Flash->success(__('The target has been saved.'));

                return $this->redirect(['action' => 'index']);
            } else {
                $this->Flash->error(__('The target could not be saved. Please, try again.'));
            }
        }
        $phases = $this->Targets->Phases->find('list', ['limit' => 200]);
        $this->set(compact('target', 'phases'));
        $this->set('_serialize', ['target']);
    }


I suspect either my code is wrong or the behavior is not tagging the entity with dirty() as the beforeSave event does not get run. If I comment out this within \ORM\Table:

        if ($entity->isNew() === false && !$entity->dirty()) {
            return $entity;
        }

the Timestamp behavior updates the 'updated' field as expected, suggesting it's related to this if statement running when it shouldn't be.

I presume the first argument checks whether the save is an edit (new === false), and the second argument must then also be true, as the routine then exits early with $entity, and not triggering the beforeSave event which follows immediately afterwards. 

I ended up looking here in the first place as I noticed that beforeSave() in my Table code does not get run at all (although, curiously, beforeMarshal() in the same function does run). 

Am I missing something very obvious here?

**Comment by markstory** [comment]
Have you ensured that the record is being saved and that there is no validation error? If there isn't then I would also ensure that you are not getting an Automodel. You can do this with `debug($this->Targets)`. If you see `Cake\ORM\Table` your table class is not being used which would be why the TimestampBehavior is not triggering.

**Comment by unclezoot** [comment]
I see the success Flash message as expected but examination of the DB shows that other altered fields get updated but not the 'updated' field I wanted (this is following the old Cake2 field naming I'm used to), so I concluded there were no validation errors here. 

I never actually saw Cake\ORM\Table but tried to work backwards from the save statement to find out why the TimeStamp does not get triggered.

From adding a skeleton beforeSave() to my my own table class then slapping  var_dump + die in different functions, seems that Targets::beforeSave never gets called, but Targets::beforeMarshal does (?)

Is there anything else I can provide to clarify what is/is not happening here?

**Comment by markstory** [comment]
@unclezoot Right now I'm not able to reproduce the issue you're having. So until we can figure that out, I don't think we'll make much progress.

**Comment by dakota** [comment]
When you save on the edit action, do you actually change any data, or just click on "Submit"?

**Comment by unclezoot** [comment]
I'm sure I changed other bits of data when testing yesterday.
Today I've just retried by editing another field and it's all working as expected (annoyingly).
I wouldn't be surprised if I hadn't changed any fields before clicking save and expected the modified field to get updated as it would in Cake2. 

I'm very sorry to have wasted your time guys. Seems I need to reread the Entity documentation a bit more closely.
